### PR TITLE
Python logging improvements

### PR DESCRIPTION
**Pull request by rcoup**
## What does this PR do?

1. Fixes an exit-time segfault when a custom error handler is installed and `CPL_DEBUG` is set.
2. Support a custom error handler via `gdal.SetErrorHandler()`. This also allows capturing logs for all  threads, which Push/Pop don't.
3. Fix some other error-handling bugs/omissions (eg. `CE_None` messages went nowhere)
4. Add a helper method, `gdal.ConfigurePythonLogging()` which:
    * sets a default GDAL error handler
    * passes all logged messages to Python's logging framework
    * turns on `CPL_DEBUG` so the log framework can decide to filter debug messages
    * maps GDAL's log levels to Python's
    * maps `CPLE_*` error codes to names

https://gist.github.com/rcoup/ba0786ead611ed21b8e24ce364e4df1f has a test script:

* `test_errhandler.py --basic` is a simple testcase for the crash
* `test_errhandler.py` will do a whole pile of tests
* `test_errhandler.py --configure` will show the `ConfigureLogging()` behaviour

## What are related issues/pull requests?

https://github.com/mapbox/rasterio/issues/902 & various others

## Tasklist

 - [ ] Review. Commits are broken out tidily.
 - [x] Assemble some of the gist into unit tests
 - [ ] Adjust for comments
 - [x] All CI builds and checks have passed

## Environment

Provide environment details, if relevant:

* OS: Linux

### CODE REVIEW

File: gdal/gcore/gdaldllmain.cpp [hunk 1]
```diff
@@ -72,8 +72,8 @@ void GDALDestroy(void)
         return;
     bGDALDestroyAlreadyCalled = true;
 
-    CPLDebug("GDAL", "In GDALDestroy - unloading GDAL shared library.");
     bInGDALGlobalDestructor = true;
+    CPLDebug("GDAL", "In GDALDestroy - unloading GDAL shared library.");
```
Comment (rouault) [inline gdal/gcore/gdaldllmain.cpp]: can you put as a code comment why order is sensitive here

File: gdal/swig/include/cpl.i [hunk 2]
```diff
@@ -50,20 +50,14 @@ typedef char retStringAndCPLFree;
     CPLDebug( msg_class, "%s", message );
   }
 
-  CPLErr SetErrorHandler( char const * pszCallbackName = NULL )
+  CPLErr SetErrorHandler( CPLErrorHandler pfnErrorHandler = NULL, void* user_data = NULL )
```
Comment (rouault) [inline gdal/swig/include/cpl.i]: so no more support for SetErrorHandler("CPLQuietErrorHandler") ?

**Comment by rcoup** [comment]
Added tests for `ConfigureLogging()` and `SetErrorHandler()` taking a callable.

**Comment by rouault** [comment]
@rcoup Are you good with that PR ? If so, you might do some commit squashing to try to compact history a bit
@sgillies Any comment ?

**Comment by sgillies** [comment]
No comments. I haven't had time to look into this enough to be useful.

**Pull request merged**